book-content/chapters/06-objects.md

```diff
@@ -0,0 +1,1386 @@
+  return obj.hasOwnProperty(key);
+};
+
+// ---cut---
+const obj = {
+  1: "bar",
+};
+```
+
+Because an object can also have a symbol as a key, there is also a test for that case. It currently has type errors for `fooSymbol` and `barSymbol` when calling `hasKey`:
+
+```ts twoslash
+// @lib: dom,es2023,dom.iterable
+// @errors: 2345
+const hasKey = (obj: object, key: string) => {
+  return obj.hasOwnProperty(key);
+};
+
+// ---cut---
+const fooSymbol = Symbol("foo");
+const barSymbol = Symbol("bar");
+
+const obj = {
+  [fooSymbol]: "bar",
+};
+```
+
+Your task is to update the `hasKey` function so that all of these tests pass. Try to be as concise as possible!
+
+<Exercise title="Exercise 4: Dynamic Key Support" filePath="/src/020-objects/086-property-key-type.problem.ts"></Exercise>
+
+#### Solution 1: Use an Index Signature for Dynamic Keys
+
+Here are the three solutions:
+
+You can use an inline index signature:
+
+```typescript
+const scores: {
+  [key: string]: number;
+} = {};
+```
+
+Or an interface:
+
+```typescript
+interface Scores {
+  [key: string]: number;
+}
+```
+
+Or a type:
+
+```typescript
+type Scores = {
+  [key: string]: number;
+};
+```
+
+Or finally, a record:
+
+```typescript
+const scores: Record<string, number> = {};
+```
+
+#### Solution 2: Default Properties with Dynamic Keys
+
+Here's how to add an index signature to the `Scores` interface to support dynamic keys along with the required keys:
+
+```typescript
+interface Scores {
+  [subject: string]: number;
+  math: number;
+  english: number;
+  science: number;
+}
+```
+
+Creating a `RequiredScores` interface and extending it looks like this:
+
+```typescript
+interface RequiredScores {
+  math: number;
+  english: number;
+  science: number;
+}
+
+interface Scores extends RequiredScores {
+  [key: string]: number;
+}
+```
+
+These two are functionally equivalent, except for the fact that you get access to the `RequiredScores` interface if you need to use that seprately.
+
+#### Solution 3: Restricting Object Keys
+
+##### A Failed First Attempt at Using Record
+
+We know that the values of the `Configurations` object will be `apiBaseUrl`, which is a string, and `timeout`, which is a number.
+
+It may be tempting to use a Record to set the key as a string and the value an object with the properties `apiBaseUrl` and `timeout`:
+
+```typescript
+type Configurations = Record<
+  string,
+  {
+    apiBaseUrl: string;
+    timeout: number;
+  }
+>;
+```
+
+However, having the key as `string` still allows for the `notAllowed` key to be added to the object. We need to make the keys dependent on the `Environment` type.
+
+##### The Correct Approach
+
+Instead, we can specify the `key` as `Environment` inside the Record:
+
+```typescript
+type Configurations = Record<
+  Environment,
+  {
+    apiBaseUrl: string;
+    timeout: number;
+  }
+>;
+```
+
+Now TypeScript will throw an error when the object includes a key that doesn't exist in `Environment`, like `notAllowed`.
```

Comment:
This is not the only place when a language like this is used but I wonder if it wouldn't be better to adjust all of them to use "report an error" over "throw an error".